File: ruby/bob/bob.rb
class Reaction
  def self.test(_stimulus)
    true
  end

  def self.response
    'Whatever.'
  end
end

class SilentReaction < Reaction
  def self.test(stimulus)
    stimulus.is_empty?
  end

  def self.response
    'Fine. Be that way!'
  end
end

class YellingReaction < Reaction
  def self.test(stimulus)
    stimulus.is_loud?
  end

  def self.response
    'Woah, chill out!'
  end
end

class QuestioningReaction < Reaction
  def self.test(stimulus)
    stimulus.is_a_question? && !stimulus.is_loud?
  end

  def self.response
    'Sure.'
  end
end

class Bob

  def initialize(reactions=[ SilentReaction,
                             YellingReaction,
                             QuestioningReaction,
                             Reaction ])
    @reactions = reactions
  end

  def hey(input)
    respond_to(Stimulus.new input)
  end

  private

  def respond_to(stimulus)
    @reactions.find { |reaction| reaction.test(stimulus) }.response
  end

end

class Stimulus
  def initialize(stimulus)
    @stimulus = stimulus.strip
  end

  def is_empty?
    @stimulus.length == 0
  end

  def is_loud?
    @stimulus.upcase == @stimulus && @stimulus.swapcase != @stimulus
  end

  def is_a_question?
    @stimulus.end_with? '?'
  end
end

File: javascript/bob/bob.js
function Bob() {

  var REACTIONS = [
    SilentReaction,
    YellingReaction,
    QuestioningReaction,
    Reaction
  ];

  function respondTo(stimulus) {
    stimulus = stimulus.trim();
    return REACTIONS.reduce(function(response, AReaction) {
      if (response) {
        return response;
      }
      var reaction = new AReaction();
      if (reaction.test(stimulus)) {
        return reaction.response;
      }
    }, null);
  }

  return {
    hey: respondTo
  };
}

function Reaction() {
  return {
    response: 'Whatever.',
    test: function always() {
      return true;
    }
  };
}

function SilentReaction() {
  return {
    response: 'Fine. Be that way!',
    test: function isSilent(input) {
      return (input === '');
    }
  }
}

function YellingReaction() {
  return {
    response: 'Woah, chill out!',
    test: function isYelling(input) {
      return (input.toUpperCase() === input && /[A-Z]/.test(input));
    }
  };
}

function QuestioningReaction() {
  return {
    response: 'Sure.',
    test: function isQuestioning(input) {
      return (input.slice(-1) === '?' && !new YellingReaction().test(input));
    }
  };
}


if (module) {
  module.exports = Bob;
}
